node/src/Pos/Util/Servant.hs

```diff
@@ -399,6 +401,22 @@ instance ( HasServer (apiType a :> LoggingApiRec config res) ctx
                     \sl -> sformat (string%": "%stext) paramName (paramVal sl)
             _ApiParamsLogInfo %~ (paramInfo :)
 
+-- | Unique identifier for request-response pair.
+newtype RequestId = RequestId Integer
+
+instance Buildable RequestId where
+    build (RequestId id) = bprint ("#"%build) id
+
+-- | We really want all servant servers to have non-overlapping ids,
+-- so singleton counter here.
+requestsCounter :: TVar Integer
+requestsCounter = unsafePerformIO $ newTVarIO 0
+{-# NOINLINE requestsCounter #-}
+
+allocRequestId :: MonadIO m => m RequestId
+allocRequestId =
+    atomically $ RequestId <$> modifyTVarS requestsCounter (identity <<+= 1)
```

Comment:
I guess this is just an observation (read bikeshedding) but the "unrolled" version is yes one line
longer but is more straightforward to understand for the uninitiated as it uses just the standard `stm` primitives:

```haskell
allocRequestId = atomically $ do
  modifyTVar' requestsCounter (+ 1)
  RequestId <$> readTVar
```

Always on the bikeshedding front, I would call this function simply `nextRequestId`.